src/libs/decoders/mp3.cpp

```diff
@@ -0,0 +1,222 @@
+    const drmp3_uint16 in_buffer_frame_capacity = 4096 / channels;
+
+    // setup our 16-bit output buffer
+    drmp3_int16* out_buffer = static_cast<drmp3_int16*>(internal->buffer);
+    drmp3_uint16 remaining_frames = (internal->buffer_size  / sizeof(drmp3_int16)) / channels;
+
+    // LOG_MSG("read: remaining_frames: %u", remaining_frames);
+    drmp3_uint16 total_samples_read = 0;
+    while (remaining_frames > 0) {
+        const drmp3_uint16 num_frames = (remaining_frames > in_buffer_frame_capacity) ? in_buffer_frame_capacity : remaining_frames;
+
+        // LOG_MSG("read-while: num_frames: %u", num_frames);
+        const drmp3_uint16 frames_just_read = static_cast<drmp3_uint16>(drmp3_read_pcm_frames_f32(p_mp3->p_dr, num_frames, in_buffer));
+
+        // LOG_MSG("read-while: frames_just_read: %u", frames_just_read);
+        if (frames_just_read == 0) break; // Reached the end.
+
+        const drmp3_uint16 samples_just_read = frames_just_read * channels;
+
+        // f32 -> s16
+        drmp3dec_f32_to_s16(in_buffer, out_buffer, samples_just_read);
+
+        remaining_frames   -= frames_just_read;
+        out_buffer         += samples_just_read;
+        total_samples_read += samples_just_read;
+    }
+    // SNDDBG(("encoded stream offset: %d", SDL_RWtell(internal->rw) ));
+
+    return total_samples_read * sizeof(drmp3_int16);
+} /* MP3_read */
+
+static Sint32 MP3_open(Sound_Sample* const sample, const char* const ext)
+{
+    Sound_SampleInternal* const internal = static_cast<Sound_SampleInternal*>(sample->opaque);
+    Sint32 result(0); // assume failure until proven otherwise
+    mp3_t* p_mp3 = (mp3_t*) SDL_calloc(1, sizeof (mp3_t));
+    if (p_mp3 != NULL) {
+        p_mp3->p_dr = (drmp3*) SDL_calloc(1, sizeof (drmp3));
+        if (p_mp3->p_dr != NULL) {
+            result = drmp3_init(p_mp3->p_dr, mp3_read, mp3_seek, sample, NULL, NULL);
+            if (result == DRMP3_TRUE) {
+                SNDDBG(("MP3: Accepting data stream.\n"));
+                sample->flags = SOUND_SAMPLEFLAG_CANSEEK;
+                sample->actual.channels = p_mp3->p_dr->channels;
+                sample->actual.rate = p_mp3->p_dr->sampleRate;
+                sample->actual.format = AUDIO_S16SYS;  // returns native byte-order based on architecture
+                const Uint64 num_frames = populate_seek_points(internal->rw, p_mp3, MP3_FAST_SEEK_FILENAME); // status will be 0 or pcm_frame_count
+                if (num_frames != 0) {
+                    const unsigned int rate = p_mp3->p_dr->sampleRate;
+                    internal->total_time = ( static_cast<Sint32>(num_frames) / rate) * 1000;
+                    internal->total_time += (num_frames % rate) * 1000 / rate;
+                    result = 1;
+                } else {
+                    internal->total_time = -1;
+                    // BAIL_MACRO("MP3: populate_seek_table failed to create seek points for the stream; falling back to brute-force seeking.", 0);
```

Comment:
These comments with BAIL_MACRO seem like no longer needed :)

Reply:
Gone!

Reply:
Fixed in https://github.com/dreamer/dosbox-staging/commit/c4b303a2414482fa5a4a02d3c094e83fba32dcf4